### DISCUSSION TITLE
Is it possible to load an internal script via <Script/> component?

### DISCUSSION THREAD

**Discussion by heymath** [vercel/next.js#36530] [Help]
Hello :wave: :smile: 

Is it possible to load an internal script via <Script src="???" /> instead of importing it at the top of the tsx file?

For instance in _app.tsx, instead of doing that:

```tsx
import { AppProps } from "next/app";

import "my-polyfill-en-US.ts";
import "my-polyfill-en-GB.ts";
import "my-polyfill-fr-FR.ts";

function MyApp(props: AppProps) {
    return (
        <div id="app">
        </div>
    );
}
```

I would like to do something like that:

```tsx
import { AppProps } from "next/app";
import Head from "next/head";
import Script from "next/script";

function MyApp(props: AppProps) {
    const { currentLocale } = useCurrentLocale();
    return (
        <div id="app">
            <Head>
                <Script src={`my-polyfill-${currentLocale}.ts`} key="my-polyfill" />
            </Head>
        </div>
    );
}
```

The problem is that Webpack build files and change their names, plus ts/tsx is converted to js files, etc.
Plus files are built into `/.next/static/**`, served from path `/_next` and it can be hosted on a CDN or path may change because it can be prefixed like we do as `/static/_next/` with the `assetPrefix` option next.config.js file.

So how can I get/guess/use the built path of a file? Is it possible? :pray:

**Comment by icyJoseph** [comment]
If these scripts are setting globals and other properties to enable further work in the application, you could, I think, create `src/scripts` , assuming your code is using `src`, and then TypeScript compile each file in scripts, and output it to the `public` folder, either keeping the result on `git` or compiling it before `build` and/or `dev`.  That way the `Script` tag could just point to `/polyfill-en.js`.

I am not sure if its possible to tell `Script` to look at relative paths, and, in addition, then make it compile the Typescript files though.

**Reply by heymath** [reply]
Hi, thanks for your answer but I am not sure to get your point :thinking: 

If I create a `src/scripts/my-script.ts` file, by default it will be built into `/.next/static/chunks/my-script-<CONTENTHASH>.js`. Plus it will be built only if it is imported as `import 'src/scripts/my-script.ts'`, but I don't want to do it. I just want to reference it as a <script> tag.

So unless I do some sorcery to change Next/Webpack config and build some files another way, it seems not possible.
And the point of using Next.js is having to do almost 0 config so I really want to avoid this kind of tricks/configs.

Also the problem with your example is if I manage to have the final file named `polyfill-en-EN.js` for instance, I will have problem with cache in case the file content change :disappointed:

**Reply by icyJoseph** [reply]
Imagine if you just wrote your scripts directly under `public`, and not import them into any page source code.

Then the scripts would be available, like a favicon, like any other static assets really, **no wepback magic at all**. 

What I am suggesting is that you isolate these scripts, by not importing them at all, and compile them before running dev/build, and place the result in public.

> You'd need to compile these files manually though, using typescript (`tsc`).

But yeah the caching would be an issue, but you could something like, https://github.com/vercel/next.js/discussions/16118, with `max-age=0`

Selectively doing polyfills like this, ain't easy. With your current solution, all polyfills are loaded. With my suggestion you gotta compile them before hand, and control cache.

---

You could use dynamic imports, but those would add a ton of complexity I think.

--- 

To answer the question at hand, I don't think it is really possible to inline a `typescript` file path to `Script`'s  source and get that compiled without implementing some kind of compilation step. At least not at the moment.

**Reply by heymath** [reply]
OK, thanks a lot for clarifying your message :ok_hand:
But yeah I would like that feature to be handled by Next.js, not having to add extra config, even less adding extra build steps / different builds :disappointed:

**Reply by EngrZaks** [reply]
should this just work normal if the script's src is a .js file?

**Comment by ydhnwb** [comment]
any possibility of this?
